http-server/datahike/http/middleware.clj

```diff
@@ -0,0 +1,93 @@
+(ns datahike.http.middleware
+  (:require
+   [buddy.auth :refer [authenticated?]]
+   [buddy.auth.backends :as buddy-auth-backends]
+   [buddy.auth.middleware :as buddy-auth-middleware]
+   [taoensso.timbre :refer [info trace]]
+   [muuntaja.core :as m]
+   [datahike.json :as json])
+  (:import
+   [clojure.lang ExceptionInfo]))
+
+(defn auth
+  "Middleware used in routes that require authentication. If request is not
+   authenticated a 401 not authorized response will be returned.
+   Dev mode always authenticates."
+  [config handler]
+  (fn [request]
+    (if (or (:dev-mode config) (authenticated? request))
+      (handler request)
+      {:status 401 :error "Not authorized"})))
+
+(defn token-auth
+  "Middleware used on routes requiring token authentication"
+  [config handler]
+  (buddy-auth-middleware/wrap-authentication
+   handler
+   (buddy-auth-backends/token {:token-name "token"
+                               :authfn     (fn [_ token]
+                                             (let [valid-token?  (not (nil? token))
+                                                   correct-auth? (= (:token config) token)]
+                                               (when (and correct-auth? valid-token?)
+                                                 "authenticated-user")))})))
+
+;; TOOD map more errors
+(defn cause->status-code [cause]
+  400)
+
+(defn wrap-server-exception [handler]
+  (fn [request]
+    (try
+      (handler request)
+      (catch ExceptionInfo e
+        (let [cause (:cause (.getData e))]
+          (info "Server error:" e (.printStackTrace e))
+          {:status (cause->status-code cause)
+           :body   {:message (.getMessage e)}})))))
+
+(defn wrap-fallback-exception
+  [handler]
+  (fn [request]
+    (try
+      (handler request)
+      (catch Exception e
+        (info "Fallback exception:" e (.printStackTrace e))
+        {:status 500
+         :body   {:message "Unexpected internal server error."}}))))
+
```

Comment:
Is there a reason why you don't use the reitit-provided exception middleware? https://github.com/metosin/reitit/blob/master/doc/ring/exceptions.md

Reply:
I can still make a 500 with a huge response ( which might lead to ddos-attacks using our server). It would be good if we can catch all errors with the provided middleware and pass a proper response. 

The error I mean is reproduced using swagger ui `database-exists`... just send it as it is provided

Reply:
I did not write this code actually, I just ported it over from datahike-server. It is good that you bring that up, I need to think about it. If you have a good approach to fix it then also please feel free to go ahead. I do not have that much everyday experience with running HTTP servers and I think it is important that datahike-server does a decent job at it (I just don't want to buy into details of the protocol that unnecessarily complicate the binding).

Reply:
https://github.com/replikativ/datahike/pull/645/files